crates/iota-transaction-builder/src/builder/mod.rs

```diff
@@ -871,23 +888,73 @@ impl<L> TransactionBuilder<Client, L> {
         Ok(res)
     }
 
-    /// Execute the transaction and optionally wait for finalization.
+    /// Execute the transaction and optionally wait for finalization. The
+    /// GraphQL client will be used unless a gas station was configured, in
+    /// which case the transaction will be sent to the endpoint for execution.
     pub async fn execute(
+        mut self,
+        keypair: &SimpleKeypair,
+        wait_for_finalization: bool,
+    ) -> Result<Option<TransactionEffects>, Error> {
+        let gas_station_data = self.data.gas_station_data.take();
+        let client = self.client.clone();
+        let mut txn = self.finish().await?;
+
+        let res = if let Some(gas_station_data) = gas_station_data {
+            let digest = gas_station_data.execute_txn(&mut txn, keypair).await?;
+            client
+                .transaction_effects(digest)
+                .await
+                .map_err(Error::Client)?
+        } else {
+            client
+                .execute_tx(
+                    &[keypair.sign_transaction(&txn).map_err(Error::Signature)?],
+                    &txn,
+                )
+                .await
+                .map_err(Error::Client)?
+        };
+
+        let mut retries_left = 100;
+        if wait_for_finalization {
+            while retries_left > 0
+                && client
+                    .transaction(txn.digest())
+                    .await
+                    .map_err(Error::Client)?
+                    .is_none()
+            {
+                tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
+                retries_left -= 1;
+            }
```

Comment:
I'm wondering if we should return an error if no retries are left
Because if the caller wants to wait for finalization, he probably wants to do something afterwards with the tx and if it doesn't exist yet, because the gas station didn't actually submit it maybe, then this could cause some issues

Reply:
true, good point